lib/client/api.go

```diff
@@ -498,28 +498,60 @@ func VirtualPathEnvNames(kind VirtualPathKind, params VirtualPathParams) []strin
 // RetryWithRelogin is a helper error handling method, attempts to relogin and
 // retry the function once.
 func RetryWithRelogin(ctx context.Context, tc *TeleportClient, fn func() error) error {
-	err := fn()
-	if err == nil {
-		return nil
+	// Don't try to re-login when using an identity file / external identity.
+	if tc.SkipLocalAuth {
+		return trace.Wrap(fn())
 	}
 
-	if utils.IsPredicateError(err) {
-		return trace.Wrap(utils.PredicateError{Err: err})
+	retryWithRelogin := func() error {
+		key, err := tc.Login(ctx)
+		if err != nil {
+			if trace.IsTrustError(err) {
+				return trace.Wrap(err, "refusing to connect to untrusted proxy %v without --insecure flag\n", tc.SSHProxyAddr)
+			}
+			return trace.Wrap(err)
+		}
+		if err := tc.ActivateKey(ctx, key); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Attempt device login. This activates a fresh key if successful.
+		if err := tc.AttemptDeviceLogin(ctx, key); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Save profile to record proxy credentials
+		if err := tc.SaveProfile(true); err != nil {
+			log.Warningf("Failed to save profile: %v", err)
+			return trace.Wrap(err)
+		}
+
+		return trace.Wrap(fn())
 	}
 
-	if !IsErrorResolvableWithRelogin(err) {
+	// First check if we have existing, unexpired credentials.
+	_, profileErr := tc.ProfileStatus()
+	if profileErr != nil {
+		log.Debugf("Activating relogin on %v.", profileErr)
+		err := retryWithRelogin()
 		return trace.Wrap(err)
 	}
 
-	// Don't try to login when using an identity file / external identity.
-	if tc.SkipLocalAuth {
-		return trace.Wrap(err)
+	fnErr := fn()
+	if fnErr == nil {
+		return nil
 	}
 
-	log.Debugf("Activating relogin on %v.", err)
+	if utils.IsPredicateError(fnErr) {
+		return trace.Wrap(utils.PredicateError{Err: fnErr})
```

Comment:
Is this "double wrapping" the predicate error? If it already is a predicate error why do we need to create a new predicate error?

Reply:
`IsPredicateError` just checks for this substring rather than looking for a `utils.PredicateError` type.

```go
// IsPredicateError determines if the error is from failing to parse predicate expression
// by checking if the error as a string contains predicate keywords.
func IsPredicateError(err error) bool {
	return strings.Contains(err.Error(), "predicate expression")
}
```

It looks like we sometimes return errors that contain "predicate expression" that we want to convert to predicate errors.

ex:
```go
return false, trace.BadParameter("failed to parse predicate expression: %s", err.Error())
```

But you're right that this could double wrap a predicate error, which is odd, but not new to this PR.